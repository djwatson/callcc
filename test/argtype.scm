(import (scheme base) (scheme write) (scheme inexact) (scheme char) (scheme complex) (scheme eval) (scheme file)
	(scheme inexact) (scheme lazy) (scheme load) (scheme process-context) (scheme read))

(define-syntax test-error
  (syntax-rules ()
    ((_ a)
     (let ((err (guard (cxp (#t #t)) a #f))) 
       (unless err (display "FAIL:") (display 'a) (newline))))))

(define-record-type foo (make-foo) foo?)

(define type-examples
  (list
   #t #\a (list 'test) '() 9739 1.0 1/2
   1000000000000000000000000000000000000000000000000000000
   -1
   1-2i
   (string-copy "test") 'test '#() (vector 'a 'b 'c) (make-foo) (lambda (x) x)))

(define-syntax test-matrix
  (syntax-rules (HERE)
    ((_ a)
     (map (lambda (type) (guard (cxp (#t #t)) (a type) #f)) type-examples))
    ((_ a HERE b)
     (map (lambda (type) (guard (cxp (#t #t)) (a type b) #f)) type-examples))
    ((_ a b HERE)
     (map (lambda (type) (guard (cxp (#t #t)) (a b type) #f)) type-examples))))

(test-error (cons 1))
(test-error (cons))
(test-error (car))
(test-error (car #f))
(test-error (car #\a))
(test-error (car 1.2))
(test-error (car car))
(test-error (car "stest"))
(test-error (car 'car))
(test-error (car #(car)))
(display (test-matrix car)) (newline)
(display (test-matrix + HERE 1)) (newline)
(display (test-matrix / HERE 1)) (newline)
(display (test-matrix abs)) (newline)
(display (test-matrix nan?)) (newline)
(display (test-matrix sin)) (newline)
(display (test-matrix asin)) (newline)
(display (test-matrix inexact)) (newline)
(display (test-matrix call/cc)) (newline)
(display (test-matrix atan)) (newline)
(display (test-matrix exp)) (newline)
(display (test-matrix log)) (newline)
(display (test-matrix finite?)) (newline)
;;(display (test-matrix set-car! HERE 1))

(define-syntax slam-test
  (syntax-rules ()
    ((_ func)
     (begin
       (display "Slam test:") (display 'func) (newline) (slam-test-go func)))))

(define debug #f)
(define (slam-test-go func)
  (map (lambda (type)
	 (when debug (display type) (newline))
	 (guard (cxp (#t #t)) (func type) #f)) type-examples)
  (map (lambda (type1)
	 (map (lambda (type2)
		(when debug
		  (display type1) (display " ")
		  (display type2) (newline))
		(guard (cxp (#t #t)) (func type1 type2) #f))
	      type-examples)) type-examples)
  (map (lambda (type1)
	 (map (lambda (type2)
		(map
		 (lambda (type3)
		   (when debug
		   (display type1) (display " ")
		   (display type2) (display " ")
		     (display type3) (newline)
		     )
		   (guard (cxp (#t #t)) (func type1 type2 type3) #f))
		 type-examples))
	      type-examples))
       type-examples))

(define-syntax slam-tests
  (syntax-rules ()
    ((_ test ...)
     (begin
       (slam-test test) ...))))

(slam-test number->string)


(slam-tests
;; base
 reverse length apply append boolean? bytevector? eof-object? complex?  real? rational? integer? exact? inexact? inexact exact char? number? procedure? string? symbol? vector? not pair? list ;
	    ;;make-list			;
	    list-set! list-copy			;
	    zero? negative? positive? odd? even? abs ;
	    boolean=? symbol=? < > <= >= + - / *	;
	    max min modulo gcd lcm			;
	    ;;	    expt			;
	    remainder quotient cdr car set-car! set-cdr! cons eq? eqv? equal? null? ;
	    string<? char=? list-ref list-tail assq member for-each   string->list
	    newline
	    vector-map vector-for-each string-map string-for-each map list->vector
	    ;;make-vector
	    vector-set! vector-ref
	    ;;make-string
	    string-fill! list? equal? vector-length string-length string-set! string-ref
	    vector->list symbol->string string->symbol
	    string->vector
	    vector->string
	    vector-copy
	    vector-fill!
	    vector-copy! vector-append string list->string number->string integer->char char->integer string-copy
	    string-copy! substring string-append vector call/cc values call-with-values dynamic-wind
	    make-parameter flush-output-port peek-char char-ready? 
	    truncate exact-integer-sqrt round rationalize floor ceiling floor/ floor-quotient floor-remainder
	    truncate/ truncate-quotient truncate-remainder imag-part real-part make-rectangular
	    square exact-integer? numerator denominator raise
	    bytevector-length bytevector-copy bytevector-u8-ref bytevector-u8-set! bytevector-copy!
	    bytevector-append utf8->string string->utf8

	    ;; char
	    char-alphabetic? char-ci<=?
		       char-ci<? char-ci=?
		       char-ci>=? char-ci>?
		       char-downcase char-foldcase
		       char-lower-case? char-numeric?
		       char-upcase char-upper-case?
		       char-whitespace? digit-value
		       string-ci<=? string-ci<?
		       string-ci=? string-ci>=?
		       string-ci>? 
		       string-downcase
		       string-foldcase string-upcase

		       ;; complex
		       angle imag-part
	    magnitude make-polar
	    make-rectangular real-part
	    ;; file
	    call-with-input-file call-with-output-file
      delete-file file-exists?
      open-binary-input-file open-binary-output-file
      open-input-file open-output-file
      with-input-from-file with-output-to-file

;;       ;;inexact
      acos asin
	   atan cos
	   exp finite?
	   infinite? log
	   nan? sin
	   sqrt tan

	   ;; lazy
	    force make-promise
	    promise?

	    ;; load
	    load
	    ;; process context
	    command-line
	    ;;emergency-exit
;;		  exit
		  get-environment-variable
		  get-environment-variables

		  ;; read
		  read

		  ;;write
		  write display write-shared write-simple
	    ;; eval
	    environment eval
	    )


