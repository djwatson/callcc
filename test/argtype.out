(#t #t #f #t #t #t #t #t #t #t #t #t #t #t #t #t)
(#t #t #t #t #f #f #f #f #f #f #t #t #t #t #t #t)
(#t #t #t #t #f #f #f #f #f #t #t #t #t #t #t #t)
(#t #t #t #t #f #f #f #f #f #t #t #t #t #t #t #t)
(#f #f #f #f #f #f #f #f #f #f #f #f #f #f #f #f)
(#t #t #t #t #f #f #f #f #f #f #t #t #t #t #t #t)
(#t #t #t #t #f #f #f #f #f #f #t #t #t #t #t #t)
(#t #t #t #t #f #f #f #f #f #f #t #t #t #t #t #t)
(#t #t #t #t #t #t #t #t #t #t #t #t #t #t #t #f)
(#t #t #t #t #f #f #f #f #f #f #t #t #t #t #t #t)
(#t #t #t #t #f #f #f #f #f #f #t #t #t #t #t #t)
(#t #t #t #t #f #f #f #f #f #f #t #t #t #t #t #t)
(#f #f #f #f #f #f #f #f #f #f #f #f #f #f #f #f)
Slam test:number->string
Slam test:reverse
Slam test:length
Slam test:apply
Slam test:append
Slam test:boolean?
Slam test:bytevector?
Slam test:eof-object?
Slam test:complex?
Slam test:real?
Slam test:rational?
Slam test:integer?
Slam test:exact?
Slam test:inexact?
Slam test:inexact
Slam test:exact
Slam test:char?
Slam test:number?
Slam test:procedure?
Slam test:string?
Slam test:symbol?
Slam test:vector?
Slam test:not
Slam test:pair?
Slam test:list
Slam test:list-set!
Slam test:list-copy
Slam test:zero?
Slam test:negative?
Slam test:positive?
Slam test:odd?
Slam test:even?
Slam test:abs
Slam test:boolean=?
Slam test:symbol=?
Slam test:<
Slam test:>
Slam test:<=
Slam test:>=
Slam test:+
Slam test:-
Slam test:/
Slam test:*
Slam test:max
Slam test:min
Slam test:modulo
Slam test:gcd
Slam test:lcm
Slam test:remainder
Slam test:quotient
Slam test:cdr
Slam test:car
Slam test:set-car!
Slam test:set-cdr!
Slam test:cons
Slam test:eq?
Slam test:eqv?
Slam test:equal?
Slam test:null?
Slam test:string<?
Slam test:char=?
Slam test:list-ref
Slam test:list-tail
Slam test:assq
Slam test:member
Slam test:for-each
Slam test:string->list
Slam test:newline
Slam test:vector-map
Slam test:vector-for-each
Slam test:string-map
Slam test:string-for-each
Slam test:map
Slam test:list->vector
Slam test:vector-set!
Slam test:vector-ref
Slam test:string-fill!
Slam test:list?
Slam test:equal?
Slam test:vector-length
Slam test:string-length
Slam test:string-set!
Slam test:string-ref
Slam test:vector->list
Slam test:symbol->string
Slam test:string->symbol
Slam test:string->vector
Slam test:vector->string
Slam test:vector-copy
Slam test:vector-fill!
Slam test:vector-copy!
Slam test:vector-append
Slam test:string
Slam test:list->string
Slam test:number->string
Slam test:integer->char
Slam test:char->integer
Slam test:string-copy
Slam test:string-copy!
Slam test:substring
Slam test:string-append
Slam test:vector
Slam test:call/cc
Slam test:values
Slam test:call-with-values
Slam test:dynamic-wind
Slam test:make-parameter
Slam test:flush-output-port
Slam test:peek-char
Slam test:char-ready?
Slam test:truncate
Slam test:exact-integer-sqrt
Slam test:round
Slam test:rationalize
Slam test:floor
Slam test:ceiling
Slam test:floor/
Slam test:floor-quotient
Slam test:floor-remainder
Slam test:truncate/
Slam test:truncate-quotient
Slam test:truncate-remainder
Slam test:imag-part
Slam test:real-part
Slam test:make-rectangular
Slam test:square
Slam test:exact-integer?
Slam test:numerator
Slam test:denominator
Slam test:raise
Slam test:bytevector-length
Slam test:bytevector-copy
Slam test:bytevector-u8-ref
Slam test:bytevector-u8-set!
Slam test:bytevector-copy!
Slam test:bytevector-append
Slam test:utf8->string
Slam test:string->utf8
Slam test:char-alphabetic?
Slam test:char-ci<=?
Slam test:char-ci<?
Slam test:char-ci=?
Slam test:char-ci>=?
Slam test:char-ci>?
Slam test:char-downcase
Slam test:char-foldcase
Slam test:char-lower-case?
Slam test:char-numeric?
Slam test:char-upcase
Slam test:char-upper-case?
Slam test:char-whitespace?
Slam test:digit-value
Slam test:string-ci<=?
Slam test:string-ci<?
Slam test:string-ci=?
Slam test:string-ci>=?
Slam test:string-ci>?
Slam test:string-downcase
Slam test:string-foldcase
Slam test:string-upcase
Slam test:angle
Slam test:imag-part
Slam test:magnitude
Slam test:make-polar
Slam test:make-rectangular
Slam test:real-part
Slam test:call-with-input-file
Slam test:call-with-output-file
Slam test:delete-file
Slam test:file-exists?
Slam test:open-binary-input-file
Slam test:open-binary-output-file
Slam test:open-input-file
Slam test:open-output-file
Slam test:with-input-from-file
Slam test:with-output-to-file
Slam test:acos
Slam test:asin
Slam test:atan
Slam test:cos
Slam test:exp
Slam test:finite?
Slam test:infinite?
Slam test:log
Slam test:nan?
Slam test:sin
Slam test:sqrt
Slam test:tan
Slam test:force
Slam test:make-promise
Slam test:promise?
Slam test:load
Slam test:command-line
Slam test:get-environment-variable
Slam test:get-environment-variables
Slam test:read
Slam test:write
#t#\a(#<closure> . #<closure>)()97391.01/21000000000000000000000000000000000000000000000000000000-11-2i"aaaa"test#()#(#<closure> #<closure> #<closure>)#<record>#<closure>Slam test:display
#ta(#<closure> . #<closure>)()97391.01/21000000000000000000000000000000000000000000000000000000-11-2iaaaatest#()#(#<closure> #<closure> #<closure>)#<record>#<closure>Slam test:write-shared
#t#\a(#<closure> . #<closure>)()97391.01/21000000000000000000000000000000000000000000000000000000-11-2i"aaaa"test#()#(#<closure> #<closure> #<closure>)#<record>#<closure>Slam test:write-simple
#t#\a(#<closure> . #<closure>)()97391.01/21000000000000000000000000000000000000000000000000000000-11-2i"aaaa"test#()#(#<closure> #<closure> #<closure>)#<record>#<closure>Slam test:environment
Slam test:eval
